src/os/bluestore/BlueStore.cc

```diff
@@ -7577,6 +7580,10 @@ void BlueStore::_kv_sync_thread()
 	_txc_applied_kv(txc);
 	--txc->osr->kv_committing_serially;
 	txc->state = TransContext::STATE_KV_SUBMITTED;
+	if (txc->osr->kv_submitted_waiters) {
```

Comment:
Doesn't it make sense to do notification after this loop completes. We can avoid excessive mutex acquisitions and  thread switches this way. flush() wouldn't proceed anyway until we complete this kv_submitting list enumeration hence no need to notify it that early.